### PR TITLE
Fix safety invariants for `WorldQuery::fetch` and simplify cloning

### PR DESCRIPTION
**Pull request by JoJoJet**
# Objective

Cloning a `WorldQuery` type's "fetch" struct was made unsafe in #5593, by adding the `unsafe fn clone_fetch` to `WorldQuery`. However, as that method's documentation explains, it is not the right place to put the safety invariant:

> While calling this method on its own cannot cause UB it is marked `unsafe` as the caller must ensure that the returned value is not used in any way that would cause two `QueryItem<Self>` for the same `archetype_index` or `table_row` to be alive at the same time.

You can clone a fetch struct all you want and it will never cause undefined behavior -- in order for something to go wrong, you need to improperly call `WorldQuery::fetch` with it (which is marked unsafe). Additionally, making it unsafe to clone a fetch struct does not even prevent undefined behavior, since there are other ways to incorrectly use a fetch struct. For example, you could just call fetch more than once for the same entity, which is not currently forbidden by any documented invariants.

## Solution

Document a safety invariant on `WorldQuery::fetch` that requires the caller to not create aliased `WorldQueryItem`s for mutable types. Remove the `clone_fetch` function, and add the bound `Fetch: Clone` instead.

---

## Changelog

- Removed the associated function `WorldQuery::clone_fetch`, and added a `Clone` bound to `WorldQuery::Fetch`.

### `fetch` invariants

The function `WorldQuery::fetch` has had the following safety invariant added:

> If this type does not implement `ReadOnlyWorldQuery`, then the caller must ensure that it is impossible for more than one `Self::Item` to exist for the same entity at any given time.

This invariant was always required for soundness, but was previously undocumented. If you called this function manually anywhere, you should check to make sure that this invariant is not violated.

### Removed `clone_fetch`

The function `WorldQuery::clone_fetch` has been removed. The associated type `WorldQuery::Fetch` now has the bound `Clone`.

Before:

```rust
struct MyFetch<'w> { ... }

unsafe impl WorldQuery for MyQuery {
    ...
    type Fetch<'w> = MyFetch<'w>
    unsafe fn clone_fetch<'w>(fetch: &Self::Fetch<'w>) -> Self::Fetch<'w> {
        MyFetch {
            field1: fetch.field1,
            field2: fetch.field2.clone(),
            ...
        }
    }
}
```

After:

```rust
#[derive(Clone)]
struct MyFetch<'w> { ... }

unsafe impl WorldQuery for MyQuery {
    ...
    type Fetch<'w> = MyFetch<'w>;
}
```

### CODE REVIEW

File: crates/bevy_ecs/src/query/fetch.rs [hunk 1]
```diff
@@ -388,6 +380,9 @@ pub unsafe trait WorldQuery {
     ///
     /// Must always be called _after_ [`WorldQuery::set_table`] or [`WorldQuery::set_archetype`]. `entity` and
     /// `table_row` must be in the range of the current table and archetype.
+    ///
+    /// If this type does not implement [`ReadOnlyWorldQuery`], then the caller must ensure that
+    /// no two `Self::Item`s may exist for the same entity at any given time.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/fetch.rs:385]:
```suggestion
    /// it is impossible for more than one `Self::Item`s to exist for the same entity at any given time.
```

**Review by alice-i-cecile — APPROVED** [review]
Good change: these safety invariants are clearer and more correct. Code simplification is a nice win too.

File: crates/bevy_ecs/src/query/fetch.rs [hunk 2]
```diff
@@ -399,6 +391,9 @@ pub unsafe trait WorldQuery {
     ///
     /// Must always be called _after_ [`WorldQuery::set_table`] or [`WorldQuery::set_archetype`]. `entity` and
     /// `table_row` must be in the range of the current table and archetype.
+    ///
+    /// If this type does not implement [`ReadOnlyWorldQuery`], then the caller must ensure that
+    /// it is impossible for more than one `Self::Item` to exist for the same entity at any given time.
```
[the remark this answers is not in the document]
Comment (JoJoJet) [reply crates/bevy_ecs/src/query/fetch.rs]: I updated the wording, do you prefer it now?

**Review by cart — APPROVED** [review]
Just did a review. I'm on board, but something weird is going on with this branch. I can't do a merge locally because the branches have "unrelated histories", nor can I rebase without getting a ton of conflicts from files untouched by this PR.

Git has a way of overriding this for merges (`--allow-unrelated-histories`) but thats the kind of thing that makes me uncomfortable.

A couple of questions for @JoJoJet :
1. Can you think of any reasons for why that would happen?
2. Can you re-apply these changes somehow on a valid branch from `main`?

**Comment by cart** [comment]
Moving to 0.12 as we've run out of time and this branch needs fixing. This is good to fix, but its not a pressing issue.

**Comment by JoJoJet** [comment]
Not sure what the problem was, but it seems to be resolved now.

**Pull request merged**